File: Client/src/app/services/hotel.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { forkJoin, map, Observable, switchMap } from 'rxjs';
import { IProvince, IAmphure, ITumbol } from '../interfaces/province.interface';
import { IHotel, IHotelCriteria } from '../pages/hotel/interfaces/hotel.interface';
import { EntityService } from './entity.service';
import { ProvinceService } from './province.service';

@Injectable({
  providedIn: 'root'
})
export class HotelService extends EntityService<number, IHotel, IHotel, IHotelCriteria> {

  override url: string = 'Hotel/';
  constructor(
    protected override http: HttpClient,
    private provinceService: ProvinceService,
  ) {
    super(http);
  }

  getData(id: number): Observable<{ data: IHotel, province: IProvince, amphure: IAmphure, tumbol: ITumbol }> {
    const obs = (data: IHotel) => {
      return {
        province: this.provinceService.getProvincesById(data.provinceId!),
        amphure: this.provinceService.getAmphureById(data.amphureId!),
        tumbol: this.provinceService.getTumbolById(data.tumbolId!),
      }
    }

    return super.getById(id)
      .pipe(switchMap((res) => forkJoin(obs(res))
        .pipe(map(({ province, amphure, tumbol }) => {
          return { data: res, province: province, amphure: amphure, tumbol: tumbol }
        }))));
  }
}

File: BL/Services/HotelService.cs
using BL.Data;
using BL.DTOs;
using BL.Entities;

namespace BL.Services
{
    internal class HotelService : EntityService<int, Hotel, ReadHotelDto, CreateUpdateHotelDto, CreateUpdateHotelDto, QueryHotelDto>, IHotelService
    {
        private readonly ProvinceService _provinceService;
        public HotelService(DataContext context, ProvinceService provinceService) : base(context)
        {
            _provinceService = provinceService;
        }

        public override ReadHotelDto Add(Hotel entity)
        {
            var province = _provinceService.GetProviceById(entity.ProvinceId);
            var amphure = _provinceService.GetAmphureById(entity.AmphureId);
            var tumbol = _provinceService.GetTumbolById(entity.TumbolId);
            var tumbolName = string.IsNullOrEmpty(tumbol!.Name) ? "" : tumbol.Name;
            var amphureName = string.IsNullOrEmpty(amphure!.Name) ? "" : amphure.Name;
            var provinceName = string.IsNullOrEmpty(province!.Name) ? "" : province.Name;
            entity.Address = $"{tumbolName} / {amphureName} / {provinceName}";
            return base.Add(entity);
        }

        public override ReadHotelDto Update(Hotel entity)
        {
            var province = _provinceService.GetProviceById(entity.ProvinceId);
            var amphure = _provinceService.GetAmphureById(entity.AmphureId);
            var tumbol = _provinceService.GetTumbolById(entity.TumbolId);
            var tumbolName = string.IsNullOrEmpty(tumbol!.Name) ? "" : tumbol.Name;
            var amphureName = string.IsNullOrEmpty(amphure!.Name) ? "" : amphure.Name;
            var provinceName = string.IsNullOrEmpty(province!.Name) ? "" : province.Name;
            entity.Address = $"{tumbolName} / {amphureName} / {provinceName}";
            return base.Update(entity);
        }
    }
}
